### DISCUSSION TITLE
Android client guide 安卓客户端游玩教程

### DISCUSSION THREAD

**Discussion by happyme531** [Melledy/Grasscutter#446] [Show and tell]
Prerequisite:  
- A server running grasscutter, maybe a computer.  
- An android phone (Optional: with Magisk)
- Running in the same LAN.  

-------

Steps *with Magisk*: 
1. Install the grasscutter server normally. Make sure the server is the most recent release or you will likely to stuck at loading screen.
2. Change both the PublicIp in config.json to your server's local network ip.  
3. Change the REMOTE_HOST value in proxy_config.py to your server's local network ip.  
4. Start mitmproxy. `mitmdump -s proxy.py -k --ssl-insecure --set block_global=false`
5. Configure proxy settings in your phone:
    - Open Android’s Settings app and tap “Wi-Fi” to view a list of Wi-Fi networks.
    - Long-press (or click the info button of) the name of the Wi-Fi network you want to change the proxy settings for. Tap “Modify Network” when a menu appears.
    - Expand the “Advanced options” section on this screen. Tap the “Proxy” option and select “Manual” to manually enter proxy server settings
    - Set hostname: <your server's local network ip>, port: 8080(if you didn't change the default value)
6. Visit http://mitm.it/ in your browser. If it says the proxy doesn't work, try to restart your phone.  
7. Download and install the mitmproxy CA cert in the regular way.
8. Install "Move Certificates" module in magisk and reboot. That way the cert will move to system cert.  
9. Start grasscutter.
10. Start game client. Only global version would work. If it is connecting to the offical server, try to stop and restart the game client.
11. Have fun! 

-------

Steps *without Magisk*

1. Patch the game client apk using [apk-mitm](https://github.com/shroudedcode/apk-mitm)
2. Follow the *with Magisk* steps and just skip step 8.

-------------

## 中文教程

要求:
- 一台运行grasscutter的服务器，或者一台电脑。
- 安卓手机(可选:有Magisk)
- 运行在同一个局域网。

-----------

*有Magisk*的步骤:
1. 正常安装grasscutter服务器。确保服务器是最新版Release否则会卡在加载界面。
2. 在config.json中把两个PublicIp都改成服务器的局域网ip。  
3. 将proxy_config.py中的REMOTE_HOST值改成服务器的局域网ip。  
4. 启动mitmproxy。`mitmdump -s proxy.py -k -ssl-insecure -set block_global=false`
5. 在您的手机中配置代理设置:
    - 打开Android的设置应用程序，点击“Wi-Fi”，查看Wi-Fi网络列表。
    - 长按(或者点信息按钮)需要修改代理设置的Wi-Fi网络名称。出现菜单时，点击“修改网络”。
    - 展开本屏幕上的“高级选项”部分。点击“代理”选项，选择“手动”，手动输入代理服务器设置
    - 设置 代理主机名:<服务器的本地网络ip>，代理端口:8080(如果你没有改变默认值)
6. 在浏览器中访问http://mitm.it/ 如果它说代理没有生效，尝试重启手机。
7. 按照正常方式下载安装mitmproxy CA证书。
8. 在magisk中安装“Move Certificates”模块并重启。这样证书就会移动到系统证书内。
9. 启动grasscutter。
10. 启动游戏客户端。只有外服才能用。如果它连上了官方服务器，尝试停止并重新启动游戏客户端。
11. 玩得开心!

------------

*没有Magisk*的步骤:

1. 使用[apk-mitm](https://github.com/shroudedcode/apk-mitm)修改游戏客户端
2. 按照*有Magisk*的步骤继续，不过跳过第8步

**Comment by minhngoc25a** [comment]
@happyme531 Can you update the guide with the non-root way?
https://github.com/shroudedcode/apk-mitm

They can use this tool to mod the apk and "ready" it for https inspection. When finished, we only need to install this (update, all data kept) over the official one (but we won't be able to play official server with this). When we want to switch to official server, download the official server's apk and overwrite it again.

**Reply by happyme531** [reply]
The patching process will modify the apk signature so it is impossible to "update" the patched client over offical one in theory.

**Reply by minhngoc25a** [reply]
So we'll have to download game data from scratch, right?

**Reply by happyme531** [reply]
Need someone testing.

**Reply by victor-lcr** [reply]
If you need to use the Chinese version of the client, you can try to modify the file name so that you don't have to download the resources again.Only the international version is used, and the principle is the same.(Sorry, my English is poor. I'm using machine translation)

**Comment by GreatXuxu** [comment]
大佬牛逼

**Comment by aquawius** [comment]
When i start mitmproxy in PUBLIC INTERNET ACCESS, there are tons of data package runs in my server, and finally server crashed, anybody have solution to solve it?            PS: I just set REMOTE_HOST using my server public IP,Nothing else has been changed

Chinese: 当我在*公网*上架设服务器并且使用上述mitmproxy脚本的时候,就会有大量的数据包涌入服务器,导致服务器直接被大量的数据包挤死机了(给人一种被DDOS的感觉),有哪位大神给点建议?    PS:我仅仅设定了REMOTE_HOST为服务器公网IP,其他的没动

**Reply by coldshineb** [reply]
I may face the same problem, I often disconnect from my proxy server after some minutes automatically, then I see mitmdump is stuck and doesn't accept connections any more

**Reply by victor-lcr** [reply]
我通过修改服务器上的代理脚本，设置将除了游戏登录数据以外的全部流量重定向到错误的web页面，这样服务器压力小了很多。
而且也减少了很多服务器的无效流量（我正在使用阿里云的按量付费实例）
你可以参考这样修改，经过我的测试没有发现问题，但是客户端必须在游戏登录后立即关闭代理，否则会无法使用网络。
```python
##
#
#   Copyright (C) 2002-2022 MlgmXyysd All Rights Reserved.
#
##

##
#
#   Genshin Impact script for mitmproxy
#
#   https://github.com/MlgmXyysd/
#
#   *Original fiddler script from https://github.lunatic.moe/fiddlerscript
#
#   Environment requirement:
#     - mitmdump from mitmproxy
#
#   @author MlgmXyysd
#   @version 1.1
#
##

from mitmproxy import http
from proxy_config import USE_SSL
from proxy_config import REMOTE_HOST
from proxy_config import REMOTE_PORT

class MlgmXyysd_Genshin_Impact_Proxy:

    LIST_DOMAINS = [
        "api-os-takumi.mihoyo.com",
        "hk4e-api-os-static.mihoyo.com",
        "hk4e-sdk-os.mihoyo.com",
        "dispatchosglobal.yuanshen.com",
        "osusadispatch.yuanshen.com",
        "account.mihoyo.com",
        "log-upload-os.mihoyo.com",
        "dispatchcntest.yuanshen.com",
        "devlog-upload.mihoyo.com",
        "webstatic.mihoyo.com",
        "log-upload.mihoyo.com",
        "hk4e-sdk.mihoyo.com",
        "api-beta-sdk.mihoyo.com",
        "api-beta-sdk-os.mihoyo.com",
        "cnbeta01dispatch.yuanshen.com",
        "dispatchcnglobal.yuanshen.com",
        "cnbeta02dispatch.yuanshen.com",
        "sdk-os-static.mihoyo.com",
        "webstatic-sea.mihoyo.com",
        "webstatic-sea.hoyoverse.com",
        "hk4e-sdk-os-static.hoyoverse.com",
        "sdk-os-static.hoyoverse.com",
        "api-account-os.hoyoverse.com",
        "hk4e-sdk-os.hoyoverse.com",
        "overseauspider.yuanshen.com",
        "gameapi-account.mihoyo.com",
        "minor-api.mihoyo.com",
        "public-data-api.mihoyo.com",
        "uspider.yuanshen.com",
        "sdk-static.mihoyo.com",
        "abtest-api-data-sg.hoyoverse.com",
        "log-upload-os.hoyoverse.com"
    ]

    def request(self, flow: http.HTTPFlow) -> None:
        if flow.request.host in self.LIST_DOMAINS:
            if USE_SSL:
                flow.request.scheme = "https"
            else:
                flow.request.scheme = "http"
            flow.request.host = REMOTE_HOST
            flow.request.port = REMOTE_PORT 
        else:
            flow.request.scheme = "http"
            flow.request.host = "127.0.0.1" # 错误web页面地址
            flow.request.port = 80

addons = [
	MlgmXyysd_Genshin_Impact_Proxy()
]

```

**Comment by jinghuashang** [comment]
不能使用fd吗?，没用过mi

**Reply by happyme531** [reply]
不能使用mi吗？，没用过fd

**Reply by jinghuashang** [reply]
听君一席话，如听一席话

**Comment by jinghuashang** [comment]
mi ca证书哪里下载哦

**Reply by zhaodice** [reply]
证书都是自己生成的

**Reply by jinghuashang** [reply]
咋生成哦，我是在termux下运行mi的

**Reply by zhaodice** [reply]
> 咋生成哦，我是在termux下运行mi的

那我可能帮不了你，我的方案是直接手机装好证书后通过与服务器建立VPN来实现私服。

**Reply by zhaodice** [reply]
至于mitmdump，我就没成功过

**Reply by jinghuashang** [reply]
> > 咋生成哦，我是在termux下运行mi的
> 
> 那我可能帮不了你，我的方案是直接手机装好证书后通过与服务器建立VPN来实现私服。

我找到了个简单的连接私服的办法，用mi真的麻烦

**Comment by freezeskies** [comment]
I can't find PublicIp in config.json. The current version is 1.7.4, what should I modify in config.json now?

我在1.7.4版本的config.json里已经找不到PublicIp了，我现在应该设置哪些？

**Comment by AndrewNg0629** [comment]
How can I download game assets on my phone? Unlike the pc version which can have the assets embedded, the android version needs to download them itself. And plus my phone is unrooted, I have no idea about where and how to put the manually downloaded resources. I am very sure that my server and mitm proxy are running well and the android client was properly patched by apk-mitm. I would appreciate it a lot if somebody could help me out.

**Comment by xiungyun** [comment]
> 先决条件：
> 
> * 运行 grasscutter 的服务器，也许是一台计算机。
> * 安卓手机（可选：带 Magisk）
> * 在同一个局域网内运行。
> 
> _使用 Magisk 的_步骤：
> 
> 1. 正常安装 Grasscutter 服务器。请确保服务器是最新版本，否则您可能会卡在加载界面。
> 2. 将 config.json 中的 PublicIp 更改为服务器的本地网络 IP。
> 3. 将 proxy_config.py 中的 REMOTE_HOST 值更改为您的服务器的本地网络 ip。
> 4. 启动 mitmproxy。`mitmdump -s proxy.py -k --ssl-insecure --set block_global=false`
> 5. 在手机中配置代理设置：
>    
>    * 打开 Android 的“设置”应用，然后点击“Wi-Fi”以查看 Wi-Fi 网络列表。
>    * 长按（或点击信息按钮）要更改代理设置的Wi-Fi网络名称。出现菜单时，点击“修改网络”。
>    * 展开此屏幕上的“高级选项”部分。点击“代理”选项，然后选择“手动”以手动输入代理服务器设置。
>    * 设置主机名：<您的服务器的本地网络 ip>，端口：8080（如果您没有更改默认值）
> 6. 在浏览器中访问http://mitm.it/。如果显示代理服务器无效，请尝试重启手机。
> 7. 以常规方式下载并安装 mitmproxy CA 证书。
> 8. 在 Magisk 中安装“移动证书”模块并重启。这样证书就会移动到系统证书中。
> 9. 启动割草机。
> 10. 启动游戏客户端。只有国际版才支持。如果正在连接到官方服务器，请尝试停止并重新启动游戏客户端。
> 11. 玩得开心！
> 
> _不使用 Magisk 的_步骤__
> 
> 1. [使用apk-mitm](https://github.com/shroudedcode/apk-mitm)修补游戏客户端 apk[](https://github.com/shroudedcode/apk-mitm)
> 2. 按照_Magisk 的_步骤操作，直接跳过步骤 8。
> 
> ## 中文教程
> 要求：
> 
> * 一台运行grasscutter的服务器，或者一台电脑。
> * 安卓手机(任选:有Magisk)
> * 在同一个楼梯上运行。
> 
> _有Magisk_的步骤：
> 
> 1. 正常安装grasscutter服务器。确保服务器是最新版本发布，否则会卡在加载界面。
> 2. 在config.json中把两个PublicIp都改成服务器的局域网ip。
> 3. 将proxy_config.py中的REMOTE_HOST值改成服务器的局域网ip。
> 4. 启动mitmproxy。`mitmdump -s proxy.py -k -ssl-insecure -set block_global=false`
> 5. 在您的手机中配置代理设置：
>    
>    * Android的设置应用程序，点击“Wi-Fi”，打开查看Wi-Fi网络列表。
>    * 长按（或者点信息按钮）需要修改代理设置的Wi-Fi网络名称。出现菜单时，点击“修改网络”。
>    * 展开本屏幕上的“高级选项”部分。点击“代理”选项，选择“手动”，手动输入代理服务器设置
>    * 设置代理主机名:<服务器的本地网络ip>，代理端口:8080(如果不改变默认值)
> 6. 在浏览器中访问http://mitm.it/如果说代理没有生效，尝试重启手机。
> 7. 按照正常方式下载安装mitmproxy CA证书。
> 8. 在magisk中安装“移动证书”模块并重新启动。这样证书就会移动到系统证书内。
> 9. 启动grasscutter。
> 10. 启动游戏客户端。只有外服才能用。如果连上了官方服务器，尝试停止并重新启动游戏客户端。
> 11. 玩得开心！
> 
> _没有Magisk_的步骤：
> 
> 1. 使用[apk-mitm](https://github.com/shroudedcode/apk-mitm)修改游戏客户端
> 2. 依次_有Magisk_的步骤继续，不过跳过第8步

纯萌新，第四步看不懂，mitmproxy是什么项目
